tests/unit/s2n_cipher_suite_match_test.c

```diff
@@ -781,6 +798,460 @@ int main(int argc, char **argv)
+                 * chacha20 cipher suite 
+                 */
+                {
+                    EXPECT_OK(s2n_toggle_chacha20_boosting(&cipher_preferences, true));
+                    cipher_preferences.allow_chacha20_boosting = true;
+
+                    EXPECT_SUCCESS(s2n_set_cipher_as_sslv2_server(connection, ecdhe_ecdsa_chacha20_boosted_wire, count));
+                    EXPECT_EQUAL(connection->secure->cipher_suite, &s2n_ecdhe_ecdsa_with_chacha20_poly1305_sha256);
+                }
+
+                /* Server has chacha20 boosting disabled. SSLv2 Server negotiates its most preferred
+                 * cipher suite, which is ecdhe_ecdsa_aes_128_cbc
+                 */
+                {
+                    EXPECT_OK(s2n_toggle_chacha20_boosting(&cipher_preferences, false));
+
+                    EXPECT_SUCCESS(s2n_set_cipher_as_sslv2_server(connection, ecdhe_ecdsa_chacha20_boosted_wire, count));
+                    EXPECT_EQUAL(connection->secure->cipher_suite, &s2n_ecdhe_ecdsa_with_aes_128_cbc_sha256);
+                }
+            }
+
+            if (!s2n_chacha20_poly1305.is_available()) {
+                /* Server and client have chacha20 boosting enabled. SSLv2 Server negotiates its most preferred
+                 * cipher suite. Because chacha20 is not available, it negotiates ecdhe_ecdsa_aes_128_cbc 
+                 */
+                {
+                    EXPECT_OK(s2n_toggle_chacha20_boosting(&cipher_preferences, true));
+
+                    EXPECT_SUCCESS(s2n_set_cipher_as_sslv2_server(connection, ecdhe_ecdsa_chacha20_boosted_wire, count));
+                    EXPECT_EQUAL(connection->secure->cipher_suite, &s2n_ecdhe_ecdsa_with_aes_128_cbc_sha256);
+                }
+            }
+        }
+
+        /* Test ChaCha20 boosting behaviour for tls 1.2 connection w/ that supports only rsa certs */ 
+        {
+            DEFER_CLEANUP(struct s2n_connection *connection = s2n_connection_new(S2N_SERVER), s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(connection);
+
+            connection->security_policy_override = &security_policy;
+            connection->kex_params.server_ecc_evp_params.negotiated_curve = s2n_all_supported_curves_list[0];
+            EXPECT_SUCCESS(s2n_connection_set_all_protocol_versions(connection, S2N_TLS12));
+
+            security_policy.minimum_protocol_version = S2N_TLS12;
+
+            DEFER_CLEANUP(struct s2n_config *rsa_only_config = s2n_config_new(), s2n_config_ptr_free);
+            EXPECT_NOT_NULL(rsa_only_config);
+            EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(rsa_only_config, rsa_cert));
+            EXPECT_SUCCESS(s2n_connection_set_config(connection, rsa_only_config));
+
+            static struct s2n_cipher_suite *cipher_suites_with_tls12_chacha20[] = {
+                &s2n_ecdhe_ecdsa_with_aes_128_cbc_sha256,       /* 0xC0,0x23 */
```

Comment:
not clear why the server has ECDSA cipher suites while it only has RSA cert?

Reply:
I purposefully added ecdsa cipher suites to test that we correctly validate the cipher suite before negotiating it. In this example, we can't negotiate `&s2n_ecdhe_ecdsa_with_aes_128_cbc_sha256`, even though it's the server's most preferred and is also available for the client. The same logic applies to the chacha20 ciphersuites.

Reply:
Thanks it makes sense.